### ISSUE TITLE
Make interface changes to allow CRAM to work with IGV

### ISSUE THREAD

**Issue by droazen** [samtools/htsjdk#597]
From @jrobinso, there are a few small adjustments we need to make to `CRAMFileReader` and  `ReferenceSource` to allow CRAM to work with IGV:

    Here’s what’s missing wrt IGV requirements:

    (1) At the time I last tries all the constructors for CRAMFileReader  took File,  rather than
    SeekableStream,  and therefore did not support http.  This might have changed
    (2) I would like to supply the ReferenceSource object from  IGV,  as IGV user already has this loaded
    and the htsjdk implementation is too memory intensive for IGV use.    Currently this is a concrete
    class,  it would be easier for me to implement and supply my own if it were an interface.  This 
    conversion (class -> interface) is fairly easy,  Intellij has tools do to this.  It looks like I would only have
     to implement one method,     
    public synchronized byte[] getReferenceBases(final SAMSequenceRecord record, final boolean tryNameVariants) {
  
    So in summary if you provided a constructor for CRAMFileReader along these lines I should be able 
    to do the rest, or at least take a first crack at it

    public CRAMFileReader(final Seekable inputStream, final SeekableStream indexInputStream,
                          final ReferenceSource referenceSource, final ValidationStringency validationStringency)

**Comment by droazen** [comment]
For @cmnbroad. Could you please prioritize this ticket over any other non-emergency fixes in htsjdk? Thanks!

**Comment by droazen** [comment]
This ticket also includes getting fixes in for https://github.com/samtools/htsjdk/issues/331 and https://github.com/samtools/htsjdk/issues/332. I believe that @jrobinso can supply the fixes for these, if we can supply unit tests for the fixes.

**Comment by yfarjoun** [comment]
I think that the PR's were closed due to lack of tests...https://github.com/samtools/htsjdk/pull/335

**Comment by jrobinso** [comment]
Yes that's correct,  but the fixes are there in the closed request.   As 
you can see its fairly minor.

**Comment by droazen** [comment]
We've agreed to implement the tests necessary to get the patches merged.

**Comment by yfarjoun** [comment]
yes. I just wanted to point out that there are fixes written that just need
tests.

**Comment by jrobinso** [comment]
I just pulled the latest htsjdk and had a look.  I think the following constructor can be used by IGV without modification:  

    public CRAMFileReader(final InputStream inputStream, final SeekableStream indexInputStream,
                          final ReferenceSource referenceSource, final ValidationStringency validationStringency) throws IOException {
 

I will be passing in SeekableStream for the inputStream.   Later, when query() is invoked,  there is an "instanceof" test on the stream in getSeekableStreamOrFailWithRTE(),  if its a SeekableStream it is used.  So I think we are good on the constructor.

**Comment by jrobinso** [comment]
So just confirming that it looks like the only requirement is to make ReferenceSource in the above constructor an interface which I will implement and supply.   And of course fixes for #331, and #332 before release,  but I don't need those to get started working.

**Comment by cmnbroad** [comment]
@jrobinso Yes - the constructor you referenced above will use a SeekableStream for queries if given one (there are [tests](https://github.com/samtools/htsjdk/blob/master/src/tests/java/htsjdk/samtools/CRAMFileBAIIndexTest.java#L70) that  exercise this).  Also, be aware that there are quite a few cases where CRAM index queries will give incorrect results at the moment, though they should be fixed soon with [this PR](https://github.com/samtools/htsjdk/pull/552).

**Comment by jrobinso** [comment]
OK, good to know,  IGV of course exclusively uses indexed queries and there's no point in adding CRAM support if they give incorrect results.

**Comment by yfarjoun** [comment]
I think that the "incorrect results" are only when there's more than one
interval. I imagine that IGV only asks for one interval at a time.

(reads currently get returned for every interval they intersect)

**Comment by jrobinso** [comment]
That's correct Yossi,  always 1 interval at a time.

**Comment by droazen** [comment]
It's worse than that @yfarjoun, I believe -- I think it can also return incorrect results in the case of unmapped reads that are assigned the positions of their mapped mates (also fixed in @cmnbroad's PR). I've pinged the reviewer of https://github.com/samtools/htsjdk/pull/552 to try to move it along.

**Comment by jrobinso** [comment]
Oh yikes,  I missed something very important.    This interface is not ideal for IGV,  and might not even work.   It fetches the reference bases for the entire sequence,  not just the interval needed for a particular query

    public synchronized byte[] getReferenceBases(final SAMSequenceRecord record,
                                                 final boolean tryNameVariants) {

IGV never loads an entire sequence.   It would be far preferable if the interface were something like

    public synchronized byte[] getReferenceBases(String chr, int start, int end,
                                                 final boolean tryNameVariants) {

I do remember now starting down this road.  The returned references are used in  the "normalize" method which has the following signature

   public void normalize(final ArrayList<CramCompressionRecord> records,
                          final byte[] ref, final int refOffset_zeroBased,
                          final SubstitutionMatrix substitutionMatrix) {

and then later in

    private static byte[] restoreReadBases(final CramCompressionRecord record, final byte[] ref,
                                           final int refOffsetZeroBased, final SubstitutionMatrix substitutionMatrix) {

So this design assumes you have lots of memory and can load entire sequences.   

Note that the reference bytes and refOffset_zeroBased are always passed together.  It might be possible to pass a partial sequence  (i.e. an interval) by changing the offset.

**Comment by jrobinso** [comment]
We could first try with the current interface for getReferenceBases,  it might be intolerably slow or require too much memory,  or it might not.

**Comment by cmnbroad** [comment]
Right - I think the original ReferenceSource was designed to target the EBI ENA web service, where the granularity is the whole sequence as identified by md5 or sha1. For now lets get the existing interface working and then we can expand if/as necessary.

**Comment by cmnbroad** [comment]
Rather than hijacking the name ReferenceSource for the interface, I left the existing concrete implementation named ReferenceSource and used the name CRAMReferenceSource for the interface since I think that will minimize source incompatibilities.
